server/audit_math/suite.py

```diff
@@ -1,718 +0,0 @@
-        (winner, loser): {"o1": o1, "o2": o2, "u1": u1, "u2": u2,}
-    }
-
-    hyp_cvr_stratum = BallotComparisonStratum(
-        cvr_stratum.num_ballots,
-        cvr_stratum.vote_totals,
-        hyp_misstatements,
-        hyp_sample_size,
-    )
-
-    # Set up the no-CVR stratum, assuming the sample looks like the
-    # prior round
-    prev_sample_size = bp_stratum.sample_size
-    hyp_sample: BallotPollingStratum.SAMPLE_RESULTS = {"hyp_round": {}}
-
-    cumulative_sample = bravo.compute_cumulative_sample(bp_stratum.sample)
-
-    # Add fake ballots to the hypothetical sample:
-    if prev_sample_size == 0:
-        # If no ballots have been sampled, assume the sample is roughly
-        # the margin
-        hyp_sample["hyp_round"] = {
-            winner: min(
-                int(n2 * (bp_stratum.vote_totals[winner] / bp_stratum.num_ballots)),
-                bp_stratum.vote_totals[winner],
-            ),
-            loser: min(
-                math.ceil(
-                    n2 * (bp_stratum.vote_totals[loser] / bp_stratum.num_ballots)
-                ),
-                bp_stratum.vote_totals[loser],
-            ),
-        }
-    else:
-        # Otherwise use the sample we've seen so far
-        hyp_sample["hyp_round"] = {
-            winner: min(
-                int(
-                    (n2 - prev_sample_size)
-                    * (cumulative_sample[winner] / prev_sample_size)
-                ),
-                bp_stratum.vote_totals[winner],
-            ),
-            loser: min(
-                math.ceil(
-                    (n2 - prev_sample_size)
-                    * (cumulative_sample[loser] / prev_sample_size)
-                ),
-                bp_stratum.vote_totals[loser],
-            ),
-        }
-
-    hyp_sample_size = n2
-    hyp_sample_size = min(bp_stratum.num_ballots, hyp_sample_size)
-
-    for rnd in bp_stratum.sample:
-        hyp_sample[rnd] = bp_stratum.sample[rnd]
-
-    hyp_no_cvr_stratum = BallotPollingStratum(
-        bp_stratum.num_ballots, bp_stratum.vote_totals, hyp_sample, hyp_sample_size
-    )
-
-    return maximize_fisher_combined_pvalue(
-        alpha, contest, hyp_no_cvr_stratum, hyp_cvr_stratum, winner, loser
-    )
-
-
-def get_sample_size(
-    risk_limit: int,
-    contest: Contest,
-    bp_stratum: BallotPollingStratum,
-    cvr_stratum: BallotComparisonStratum,
-) -> HybridPair:
-    """
-    Estimate the initial sample sizes for the audit.
-
-    Inputs:
-        risk_limit      - the risk limit (as an integer percentage) for this audit
-        contest         - the overall contest information
-        bp_stratum: a stratum object containing the ballot polling stratum
-        cvr_stratum: a stratum object containing the ballot comparison stratum
-    Outputs:
-        sample_sizes    - A Tuple of (cvr_strata_size, no_cvr_strata_size).
-    """
-
-    alpha = float(risk_limit) / 100
-
-    for winner, loser in product(contest.winners, contest.losers):
```

Comment:
i'm not sure i understand how this approach works. it seems like you perform the search for the right sample sizes for each winner/loser pair and then return the sample sizes for whichever winner/loser pair was tested last.

i.e. i think could refactor this function to look like this:

```python
def find_sample_sizes(winner, loser):
  ... everything that happens in the body of the for loop ...

for winner, loser in product(winners, losers):
  sample_sizes = find_sample_size(winner, loser)
return sample_sizes
```

it seems like the iterating over winner/loser pairs should happen within the p-value calculation when trying a sample size. something more like this:

```python
def hypothetical_pvalue(candidate_sample_size):
  hypothetical_strata = ... set up strata ...
  return compute_risk(hypothetical_strata) # reuse function that already knows how to compute risk across winner/loser pairs

candidate_sample_size = num_sampled
candidate_pvalue = 1
while candidate_pvalue > risk_limit:
  candidate_sample_size = ... some next sample size to try ...
  candidate_pvalue = hypothetical_pvalue(candidate_sample_size)
  
return candidate_sample_size
```

Reply:
Ope, that's a bug. I think what you're proposing is the right approach.